subsquid/src/dbHelper.ts

```diff
@@ -529,78 +524,87 @@ export async function getCurrentAssetPrices(ctx: Context | EntityManager): Promi
   const isRepository = ctx instanceof EntityManager;
 
   const now = new Date();
-  // Round time to the nearest minute
+  // Round time to the nearest minute.
   const date = new Date(now.getFullYear(), now.getMonth(), now.getDate(), now.getHours(), now.getMinutes());
 
+  // Current prices, that can be obtained from DB and/or Coingecko.
   const currentPrices: { [K in AssetId]?: number } = {};
-
-  let missingPrices = false;
-
-  // Check if all prices are in DB
-  for (const { assetId } of assetList.values()) {
-    try {
-      // Search for price in DB
-      const where = {
-        assetId,
-        timestamp: date
-      };
-      const assetPrice = isRepository
-        ? await ctx.getRepository(HistoricalAssetPrice).findOne({ where })
-        : await ctx.store.findOne(HistoricalAssetPrice, { where });
-
-      // If any price is missing, flag it and break
-      if (assetPrice?.price === undefined) {
-        missingPrices = true;
-        break;
-      } else {
-        currentPrices[assetId] = assetPrice.price;
-      }
-    } catch {
-      missingPrices = true;
-      break;
+  // Prices that need to be stored. Use Set to avoid duplications when multiple
+  // assets rely on the same base asset.
+  const updatePrices: Set<AssetId> = new Set();
+  // New HistoricalAssetPrices to be stored.
+  const newHistoricalPrices: Array<HistoricalAssetPrice> = [];
+
+  // Check if all prices are in DB, and flag the missing ones for updating
+  // This includes the ones that are required to calculate the spot price for
+  // another asset.
+  for (const { assetId, spotPriceBaseAsset } of assetList.values()) {
+    // Use the asset ID of the base asset if it is provided, as this asset relies
+    // on another price to calculate its own.
+    const assetIdToUse = spotPriceBaseAsset?.assetId || assetId;
+    // Search for price in DB.
+    const price = await findHistoricalAssetPrice(ctx, assetIdToUse, date);
+
+    if (price) {
+      // If price is found, keep it for returning
+      currentPrices[assetIdToUse] = price;
+    } else {
+      // If price is not found, flag it for updating
+      updatePrices.add(assetIdToUse);
     }
   }
 
-  // If any price is missing, get all from Coingecko and update DB
-  if (missingPrices) {
-    const queryIds = assetList
-      .map(({ coingeckoId }) => coingeckoId)
+  // If there are assets for updating, query Coingecko
+  if (updatePrices.size > 0) {
+    // Create query url
+    const queryIds = Array.from(updatePrices)
```

Comment:
Is Array.from needed here? Imo updatePrices.map is enough

Reply:
It's a `Set` (to avoid duplications) so it does not have the `.map` method

Reply:
Oh right.

Reply:
A valid alternative is `[...updatePrices].map` but it's essentially the same and maybe more confusing